nym-vpn-core/crates/nym-vpn-lib-types/src/tunnel_state.rs

```diff
@@ -198,6 +205,117 @@ pub enum ErrorStateReason {
     Internal(String),
 }
 
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum ClientErrorReason {
+    Firewall,
+    Routing,
+    SameEntryAndExitGateway,
+    InvalidEntryGatewayCountry,
+    InvalidExitGatewayCountry,
+    MaxDevicesReached,
+    BandwidthExceeded,
+    SubscriptionExpired,
+    Dns(String),
+    Api(String),
+    Internal(String),
+}
+
+impl From<ErrorStateReason> for ClientErrorReason {
+    fn from(value: ErrorStateReason) -> Self {
+        match value {
+            ErrorStateReason::SameEntryAndExitGateway => Self::SameEntryAndExitGateway,
+            ErrorStateReason::InvalidEntryGatewayCountry => Self::InvalidEntryGatewayCountry,
+            ErrorStateReason::InvalidExitGatewayCountry => Self::InvalidExitGatewayCountry,
+            ErrorStateReason::BadBandwidthIncrease => Self::Api(value.to_string()),
+            ErrorStateReason::SyncAccount(err) => Self::Api(err.to_string()),
+            ErrorStateReason::SyncDevice(err) => Self::Api(err.to_string()),
+            ErrorStateReason::RegisterDevice(err) => {
+                if err
+                    .message_id()
+                    .is_some_and(|id| id.contains(MAX_DEVICES_REACHED_MESSAGE_ID))
+                {
+                    Self::MaxDevicesReached
+                } else {
+                    Self::Api(err.to_string())
+                }
+            }
+            ErrorStateReason::RequestZkNym(err) => match err {
+                RequestZkNymErrorReason::VpnApi(e) => match e.message_id.as_ref() {
+                    Some(id) if id.contains(BANDWIDTH_LIMIT_REACHED_MESSAGE_ID) => {
+                        Self::BandwidthExceeded
+                    }
+                    Some(id) if id.contains(SUBSCRIPTION_EXPIRED_MESSAGE_ID) => {
+                        Self::SubscriptionExpired
+                    }
+                    _ => Self::Api(e.message),
+                },
+                RequestZkNymErrorReason::UnexpectedVpnApiResponse(message) => Self::Api(message),
+                reason => Self::Internal(reason.to_string()),
+            },
```

Comment:
suggestion: for better code clarity, thoses big pattern matching blocks could be refactored by implementing the corresponding `From`

```rust
impl From<ErrorStateReason::RequestZkNym> for ClientErrorReason {
    fn from(error: ErrorStateReason::RequestZkNym) -> Self { /* pattern matching */}
}

// repeat for ErrorStateReason::RegisterDevice, ErrorStateReason::RequestZkNymBundle etc
```

Reply:
You can take example on how it is currently implemented in tauri
https://github.com/nymtech/nym-vpn-client/blob/73086d6510ebd5a3cf2ecc1b927e9cb8cb7c7f01/nym-vpn-app/src-tauri/src/grpc/tunnel_error.rs#L54
https://github.com/nymtech/nym-vpn-client/blob/73086d6510ebd5a3cf2ecc1b927e9cb8cb7c7f01/nym-vpn-app/src-tauri/src/grpc/tunnel_error.rs#L106